### PR TITLE
Feat/binance perpetual derivative

### PR DESCRIPTION
**Pull request by yancong001**
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:

This PR create Binance connector to the latest standards.

**Tests performed by the developer**:

* Test cases for the new code

**Tips for QA testing**:
Normal QA test for Perpetuals connectors.

### CODE REVIEW

**Comment by nikspz** [comment]
Test performed: 
- Cloned and installed feature branch
- Manually created docker image successfully
- Started the client
 - Connected the connector API keys = OK
 - Balance output = OK
 - Create/Import/Start a strategy = OK
 - Check status and status --live match with exchange = OK
 - Order cancellation = OK
 - Ticket and Order book --live match with exchange = OK
 - Order filled events = OK
 - Vertify fee calculation = OK
 - Data collected by datadog = OK
 - Test different token pairs = OK
 - Position retrieval = OK
 
 Pending tests:
 - Funding payment handling
 - Overnight test, checking issues during the longrun:
    - Network reconnection, Order failed submit, Unexpected errors or Warnings

**Comment by nikspz** [comment]
Hi, @yancong001. Could you please check failing tests?

**Comment by yancong001** [comment]
> Hi, @yancong001. Could you please check failing tests?

Of course, but what are the failed tests?

**Comment by nikspz** [comment]
@yancong001 Please check tests failing here:
https://github.com/hummingbot/hummingbot/actions/runs/3757406022/jobs/6426949434

**Comment by yancong001** [comment]
> @yancong001 Please check tests failing here: https://github.com/hummingbot/hummingbot/actions/runs/3757406022/jobs/6426949434

all tests passed

**Comment by nikspz** [comment]
@yancong001 That’s great, longrun tests left, upon checking should be good. 
Then going to review from our devs.

File: hummingbot/connector/derivative/binance_perpetual/binance_perpetual_user_stream_data_source.py [hunk 1]
```diff
@@ -1,53 +1,41 @@
...
-            auth=auth,
-        )
-        self._ws_assistant: Optional[WSAssistant] = None
+        self._api_factory = api_factory
+        self._auth = auth
+        self._ws_assistants: List[WSAssistant] = []
+        self._connector = connector
         self._current_listen_key = None
         self._listen_for_user_stream_task = None
         self._last_listen_key_ping_ts = None
```
Comment (cardosofede) [inline hummingbot/connector/derivative/binance_perpetual/binance_perpetual_user_stream_data_source.py:41]: Same here, you should be using the listen for user stream of the super class

**Review by cardosofede — CHANGES_REQUESTED** [review]
The overall is okey, but the idea of this refactor is to adapt the code to the latest standards. Good job, and after this changes I will do the last review and we are ready to merge the PR!

**Comment by yancong001** [comment]
> The overall is okey, but the idea of this refactor is to adapt the code to the latest standards. Good job, and after this changes I will do the last review and we are ready to merge the PR!
Ok, I will recheck and revise it soon

**Comment by yancong001** [comment]
> The overall is okey, but the idea of this refactor is to adapt the code to the latest standards. Good job, and after this changes I will do the last review and we are ready to merge the PR!

Ok, I will recheck and revise it soon

**Comment by yancong001** [comment]
those changes have been updated. @cardosofede

**Comment by cardosofede** [comment]
Seems like you made all the requested changes, good job!

I have this issue when trying to change the leverage: OSError: Error executing request POST /leverage. HTTP status is 503. Error: {"timestamp":1673025015923,"path":"/fapi/v1/leverage","msg":"Unknown error, please check your request or try again later."}
2023-01-06 14:10:16,406 - 2925 - hummingbot.core.utils.async_utils - ERROR - Unhandled error in background task: Error executing request POST /leverage. HTTP status is 503. Error: {"timestamp":1673025016447,"path":"/fapi/v1/leverage","msg":"Unknown error, please check your request or try again later."}
Traceback (most recent call last):
  File "/Users/dardonacci/Documents/work/hummingbot/hummingbot/core/utils/async_utils.py", line 9, in safe_wrapper
    return await c
  File "/Users/dardonacci/Documents/work/hummingbot/hummingbot/connector/perpetual_derivative_py_base.py", line 356, in _execute_set_leverage
    success, msg = await self._set_trading_pair_leverage(trading_pair, leverage)
  File "/Users/dardonacci/Documents/work/hummingbot/hummingbot/connector/derivative/binance_perpetual/binance_perpetual_derivative.py", line 733, in _set_trading_pair_leverage
    set_leverage = await self._api_request(
  File "/Users/dardonacci/Documents/work/hummingbot/hummingbot/connector/derivative/binance_perpetual/binance_perpetual_derivative.py", line 816, in _api_request
    raise IOError(f"Error executing request {method.name} {path_url}. "
OSError: Error executing request POST /leverage. HTTP status is 503. Error: {"timestamp":1673025016447,"path":"/fapi/v1/leverage","msg":"Unknown error, please check your request or try again later."}
2023-01-06 14:10:16,984 - 2925 - hummingbot.core.utils.async_utils - ERROR - Unhandled error in background task: Error executing request POST /leverage. HTTP status is 503. Error: {"timestamp":1673025017032,"path":"/fapi/v1/leverage","msg":"Unknown error, please check your request or try again later."}
Traceback (most recent call last):
  File "/Users/dardonacci/Documents/work/hummingbot/hummingbot/core/utils/async_utils.py", line 9, in safe_wrapper
    return await c
  File "/Users/dardonacci/Documents/work/hummingbot/hummingbot/connector/perpetual_derivative_py_base.py", line 356, in _execute_set_leverage
    success, msg = await self._set_trading_pair_leverage(trading_pair, leverage)
  File "/Users/dardonacci/Documents/work/hummingbot/hummingbot/connector/derivative/binance_perpetual/binance_perpetual_derivative.py", line 733, in _set_trading_pair_leverage
    set_leverage = await self._api_request(
  File "/Users/dardonacci/Documents/work/hummingbot/hummingbot/connector/derivative/binance_perpetual/binance_perpetual_derivative.py", line 816, in _api_request
    raise IOError(f"Error executing request {method.name} {path_url}. "
OSError: Error executing request POST /leverage. HTTP status is 503. Error: {"timestamp":1673025017032,"path":"/fapi/v1/leverage","msg":"Unknown error, please check your request or try again later."}
2023-01-06 14:10:17,716 - 2925 - hummingbot.client.hummingbot_application - INFO - stop command initiated.

**Review by cardosofede — APPROVED** [review]
LGTM! @rapcmia let's test in a long run with the latest changes

**Review by nikspz — APPROVED** [review]
Test performed:

Cloned and installed feature branch
Manually created docker image successfully
Started the client
Connected the connector API keys = OK
Balance output = OK
Create/Import/Start a strategy = OK
Check status and status --live match with exchange = OK
Order cancellation = OK
Ticket and Order book --live match with exchange = OK
Order filled events = OK
Vertify fee calculation = OK
Data collected by datadog = OK
Test different token pairs = OK
Position retrieval = OK
Funding payment handling = OK
Overnight test, no issues during the longrun:
no issues on Network reconnection, Order failed, Unexpected errors or Warnings

**Comment by rapcmia** [comment]
Merged PR in behalf of foundation team. The updated binance perpetual will be applied on the development branch and the next Hummingbot version 1.12.0

**Pull request merged**